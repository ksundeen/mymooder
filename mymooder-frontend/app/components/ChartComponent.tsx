import { StyleSheet, View } from 'react-native';
import React, { useState, useEffect, useMemo } from 'react';
import { Countries } from '@/assets/data/Countries';
import { favoritePlaceData } from '@/assets/data/favorite-places';
import Svg, { G, Path, Circle } from "react-native-svg";
import * as d3 from 'd3';
import { Colors } from '@/app/constants/Colors';
import GeoJSON from 'geojson'
import ButtonComponent from './ButtonComponent';

type PropertiesType = {
    index: number,
    name: string,
    datetime: string,
    calmness_score: number,
    happy_score: number,
    people: string[],
    activities: string[],
    personal_weather_rating: string[],
    api_weather_rating: string[],
    api_weather_temperature: number,
    notes: string
};
type GeometryType = {
    type: GeoJSON.Point
    coordinates: [GeoJSON.Position, GeoJSON.Position]
};
type FeaturesType = {
    type: GeoJSON.Feature,
    geometry: GeometryType
    properties: PropertiesType
};
type FeatureCollectionType = {
    type: GeoJSON.FeatureCollection,
    features: FeaturesType[]
};

export function ChartComponent(props: {width: number, height: number}) {
    const { width, height } = props;

    const [countryList, setCountryList] = useState<string[]>([]);
    const [happyPointList, setHappyPointList] = useState<JSX.Element[]>([]);
    const [happyPointsVisible, setHappyPointsVisibility] = useState(true);

    let moodPointsGeojson: any = favoritePlaceData

    const mapExtent = useMemo(() => {
      return width > height / 2
         ? height / 2 
         : width;
    }, [width, height]);

    // const maxHappyScoreY = useMemo(() => {
    //     return d3.max(moodPointsGeojson.features, (moodValuesFeature: FeaturesType) =>
    //         d3.max(moodValuesFeature.properties, (d: PropertiesType) => d["happy_score"])
    //     );
    // }, [moodPointsGeojson.features]);

    const colorize = useMemo(() => {
        // const colorScale = d3.scaleSequentialSymlog(d3.interpolateReds)
        //     .domain([0, maxHappyScoreY]);
            
        const colorScale = d3.scaleLinear(["red", "blue"])
            // .domain([0, maxHappyScoreY])
            .domain([0, 10])
            .clamp(true)

        return colorScale;
    }, [])

    const toggleHappyPoints = () => {
        setHappyPointsVisibility(!happyPointsVisible);
    }

    // Retrieve only happy_score mood values
    // const allHappyScoreValues = MoodPoints.map((x: MoodTypeFeature["properties"]) => x.happy_score)

    // Original Countries using Azimuthal Equal Area projection
    const countryPaths = useMemo(() => {
        // const projection = d3.geoNaturalEarth1()
        const projection = d3.geoAzimuthalEqualArea()
            .rotate([0, -90])
            .clipAngle(150)
            .fitSize([mapExtent,mapExtent], { type: "FeatureCollection", features: Countries})
            .translate([width / 2, mapExtent / 2]);
    
        const geoPath = d3.geoPath().projection(projection);

        const svgPaths = Countries.map(geoPath);

        return svgPaths;
    }, [props.width, props.height]);

    const moodPointPaths: string[] = useMemo(() => {
        // console.log('Data: ' + moodPointsGeojson)        
        // console.log('MoodPointsGeojson.features: ' + JSON.stringify(moodPointsGeojson));

        // Mood Value Mood Points
        const projection = d3.geoNaturalEarth1()
        // const projection = d3.geoAzimuthalEqualArea()
            // .rotate([0, -90])
            // .clipAngle(150)
            .fitSize([mapExtent,mapExtent], { type: "FeatureCollection", features: moodPointsGeojson.features})
            .translate([width / 2, mapExtent / 2]);

        const geoPath = d3.geoPath().projection(projection);
        const features: any = moodPointsGeojson.features
        const svgPaths: string = features.map(geoPath);
        // console.log(`features: ${JSON.stringify(features)}`);
        // console.log(svgPaths);
        return svgPaths;
    }, [props.width, props.height]);

    // Return paths for mood points
    useEffect(() => {
        const newHappyPointList: JSX.Element[] = moodPointPaths.map((path: string, i: number) => {
            const curHappyScore = moodPointsGeojson.features[i].properties.happy_score;
            const curIndex = moodPointsGeojson.features[i].properties.index;
            // const isDataAvailable = curHappyScore.some(data: any => data.datetime === date);
            return (
                <Path
                    key={curIndex}
                    d={path}
                    // stroke={Colors.lightBlue}
                    strokeOpacity={0.3}
                    strokeWidth={0.6}
                    // strokeWidth={curHappyScore}
                    // fill={colorize(curHappyScore[dateIndex])}
                    fill={colorize(curHappyScore)}
                    // fill={Colors.lightGrey}
                    opacity={0.4} />
            );
        });

        setHappyPointList(newHappyPointList);
    }, [props.width, props.height]); 

    // Return paths for countries
    useEffect(() => {
        setCountryList(
            countryPaths.map((path: string, i: number) => {
                return (
                <Path
                    key={Countries[i].properties.name}
                    d={path}
                    stroke={Colors.lightGrey}
                    strokeOpacity={0.3}
                    strokeWidth={0.6}
                    fill={Colors.lightGrey}
                    opacity={0.4} 
                />
                )
            }));
    }, []); 

    return (
        // <View style={styles.container}>
            <Svg
                width={width}
                height={height / 2}
            >
                {/* <G> */}
                {/* Background world fill */}
                {/* <Circle
                    cx={width / 2}
                    cy={mapExtent / 2}
                    r={mapExtent / 2}
                    fill={"#3b454f"}
                /> */}
                {/* World shapes */}
                {/* {countryList.map(x => x)}
                {happyPointsVisible ? happyPointList.map(x => x): ''} */}
                {/* </G> */}
            </Svg>
            // {/* <ButtonComponent 
            //     buttonStyle={{opacity: 0.5}}
            //     onPress={toggleHappyPoints}
            //     text={"Toggle Points"}
            // ></ButtonComponent>  */}
        // </View>
    );
};

export default ChartComponent;

const styles = StyleSheet.create({
    container: {
       flex: 1,
       alignItems: 'center',
       justifyContent: 'center',
    //    backgroundColor: Colors.lightBlue
    }
 });