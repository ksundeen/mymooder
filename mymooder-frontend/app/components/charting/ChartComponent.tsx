import { Dimensions, StyleSheet, SafeAreaView } from 'react-native';
import Animated from 'react-native-reanimated';
import { NumberChart } from '@/app/database/types';
import ButtonComponent from '../ButtonComponent';
import LineChartComponent from './LineChartComponent';

const { width, height } = Dimensions.get('screen');

const CARD_WIDTH = width - (width/16);
const CARD_HEIGHT = (height / 4) + (height / 175);

const ChartComponent = ({allGraphData, refreshMoodDataCaller}: 
  {allGraphData: NumberChart[], refreshMoodDataCaller: Function}) => {

  return (
    <SafeAreaView style={styles.container}>
      {/* <Collapsible title={`${allGraphData[0].name}}` }> */}
      <Animated.View style={styles.graphCard}>
        <LineChartComponent
          graphData={allGraphData[0]}
          />
      </Animated.View>
      {/* </Collapsible> */}
      <Animated.View style={styles.graphCard}>
        <LineChartComponent
          graphData={allGraphData[1]}
        />
      </Animated.View>
      <Animated.View style={styles.graphCard}>
        <LineChartComponent
          graphData={allGraphData[2]}
        />
        </Animated.View>
        <ButtonComponent 
          extraStyles={{position: 'relative', 'marginTop': 10}}
          text={"Refresh Data"} 
          diffPadding={2}
          diffFlex={0.6}
          buttonWidth={100} 
          onPress={() => refreshMoodDataCaller()}>
        </ButtonComponent>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginTop: 5,
    alignItems: 'center',
  },
  graphCard: {
    marginBottom: 2,  
    marginTop: 2,
    elevation: 5,
    borderRadius: 20,
    height: CARD_HEIGHT,
    width: CARD_WIDTH,
    backgroundColor: 'darkgray',
    shadowColor: '#000',
    shadowOffset: {
      width: 2,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
  },
});

export default ChartComponent;