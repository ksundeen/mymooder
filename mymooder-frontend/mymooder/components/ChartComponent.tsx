import { StyleSheet, Text, View } from 'react-native';
import React, { useState, useEffect, useMemo } from 'react';
import { COUNTRIES } from '@/constants/CountryShapes';
import Svg, { G, Path, Circle } from "react-native-svg";
import * as d3 from "d3";
import { ScaledSize } from 'react-native';

export function ChartComponent(props: { dimensions: ScaledSize; }) {
    const { dimensions } = props;

    const [countryList, setCountryList] = useState([]);
    const mapExtent = useMemo(() => {
      return dimensions.width > dimensions.height / 2
         ? dimensions.height / 2 
         : dimensions.width;
   }, [dimensions]);

    const countryPaths = useMemo(() => {
        const projection = d3.geoAzimuthalEqualArea()
            .rotate([0, -90])
            .clipAngle(150)
            .fitSize([mapExtent,mapExtent], { type: "FeatureCollection", features: COUNTRIES})
            .translate([dimensions.width / 2, mapExtent / 2]);
    const geoPath = d3.geoPath().projection(projection);
    const svgPaths = COUNTRIES.map(geoPath);
    return svgPaths;
    }, [dimensions]);

    useEffect(() => {
        setCountryList(
            countryPaths.map((path: string | undefined,i: string | number) => {
                return (
                <Path
                    key={COUNTRIES[i].properties.name}
                    d={path}
                    stroke={"#aaa"}
                    strokeOpacity={0.3}
                    strokeWidth={0.6}
                    fill={"#aaa"}
                    opacity={0.4} 
                />
                )
            }));
    }, []);

    return (
        <View>
            <Svg
                width={dimensions.width}
                height={dimensions.height / 2}
            >
                <G>
                <Circle
                    cx={dimensions.width / 2}
                    cy={mapExtent / 2}
                    r={mapExtent / 2}
                    fill={"#3b454f"}
                />
                {COUNTRIES.map((x: any) => x)}
                </G>
            </Svg>=
        </View>
    );
};

export default ChartComponent;