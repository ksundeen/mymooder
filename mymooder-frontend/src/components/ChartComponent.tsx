import { StyleSheet, Text, View } from 'react-native';
import React, { useState, useEffect, useMemo } from 'react';
import { COUNTRIES } from '@/assets/data/Countries';
import Svg, { G, Path, Circle } from "react-native-svg";
import * as d3 from 'd3';
import { ScaledSize } from 'react-native';


export function ChartComponent(props: any) {
    // const { dimensions } = props;

    const [countryList, setCountryList] = useState([]);
    
    const mapExtent = useMemo(() => {
      return props.dimensions.width > props.dimensions.height / 2
         ? props.dimensions.height / 2 
         : props.dimensions.width;
   }, [props.dimensions]);

    const countryPaths = useMemo(() => {
        const projection = d3.geoAzimuthalEqualArea()
            .rotate([0, -90])
            .clipAngle(150)
            .fitSize([mapExtent,mapExtent], { type: "FeatureCollection", features: COUNTRIES})
            .translate([props.dimensions.width / 2, mapExtent / 2]);
    
            const geoPath = d3.geoPath().projection(projection);
    
            const svgPaths = COUNTRIES.map(geoPath);
    
            return svgPaths;
    }, [props.dimensions]);

    useEffect(() => {
        setCountryList(
            countryPaths.map((path: string | undefined,i: string | number) => {
                return (
                <Path
                    key={COUNTRIES[i].properties.name}
                    d={path}
                    stroke={"#aaa"}
                    strokeOpacity={0.3}
                    strokeWidth={0.6}
                    fill={"#aaa"}
                    opacity={0.4} 
                />
                )
            }));
    }, []);

    return (
        <View>
            <Svg
                width={props.dimensions.width}
                height={props.dimensions.height / 2}
            >
                <G>
                <Circle
                    cx={props.dimensions.width / 2}
                    cy={mapExtent / 2}
                    r={mapExtent / 2}
                    fill={"#3b454f"}
                />
                {countryList.map(x => x)}
                </G>
            </Svg>
        </View>
    );
};

export default ChartComponent;
