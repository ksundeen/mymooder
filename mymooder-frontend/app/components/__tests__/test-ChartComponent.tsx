import * as React from 'react';
import { render } from '@testing-library/react-native'
import { ScaledSize } from 'react-native';

import { ChartComponent } from '../ChartComponent';

it(`renders correctly`, () => {
  const size: ScaledSize = {
     height: 500, width: 200, scale: 1, fontScale: 1
  }
  const tree = render(<ChartComponent width={size.width} height={size.height}></ChartComponent>).toJSON();

  expect(tree).toMatchSnapshot();
});
