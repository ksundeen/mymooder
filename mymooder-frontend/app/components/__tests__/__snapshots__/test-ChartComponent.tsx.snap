// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RNSVGSvgView
  bbHeight={250}
  bbWidth={200}
  focusable={false}
  height={250}
  style={
    [
      {
        "backgroundColor": "transparent",
        "borderWidth": 0,
      },
      {
        "flex": 0,
        "height": 250,
        "width": 200,
      },
    ]
  }
  width={200}
>
  <RNSVGGroup
    fill={
      {
        "payload": 4278190080,
        "type": 0,
      }
    }
  />
</RNSVGSvgView>
`;
